functions/src/stories.ts

```diff
@@ -11,83 +11,161 @@ const index = algoliasearch(
-        const translation = await db
-            .doc(`stories/${storyID}/translations/${translationID}`)
-            .get()
-            .then((doc) => doc.data()!);
+        await index.partialUpdateObject(
+            {objectID: translationID, ...entry},
+            {createIfNotExists: true}
+        );
+      } else {
         await index.saveObject(
-        {
-          storyID,
-          storyAuthorID: story.metaData.authorId,
-          translationID,
-          translationAuthorID: translation.metaData.authorId,
-          language: translation.language,
-          title,
-          description,
-          tags,
-          likes: translation.metaData.likes,
-        } as storyRecord,
-        {autoGenerateObjectIDIfNotExist: true}
+            entry,
+            {autoGenerateObjectIDIfNotExist: true}
         );
       }
     });
 
-// TODO
+export const trackStoryUpdateTime = functions
+    .region("europe-central2")
+    .firestore.document(
+        "stories/{storyID}/translations/{translationID}/assets/{assetID}"
+    )
+    .onWrite(async (change, context) => {
+      if (!change.after.exists) return;
+      await storyDoc(context.params.storyID, context.params.translationID)
+          .update({
+            "metaData.lastUpdateAt": firestore.FieldValue.serverTimestamp(),
+          });
+    });
+
+export const updateStoryMeta = functions
+    .region("europe-central2")
+    .firestore.document(
+        "stories/{storyID}/translations/{translationID}"
+    )
+    .onWrite(async (change, context) => {
+      if (change.after.exists) {
+        const meta = change.after.data()?.metaData;
+        await index.partialUpdateObject({
+          objectID: context.params.translationID,
+          likes: meta.likes ?? 0,
+          views: meta.views ?? 0,
+          lastUpdateAt: meta.lastUpdateAt,
+          imageUrl: meta.imageUrl,
+        });
+      }
+    });
+
+export const indexTrendingStories = functions
+    .region("europe-central2")
+    .pubsub.schedule("every 72 hours")
+    .onRun(async () => {
+      await index.browseObjects<StoryRecord>({
+        query: "",
+        attributesToRetrieve: ["storyID"],
+        batch: async (batch) => {
+          for (const hit of batch) {
+            const {storyID, objectID} = hit;
+            const doc = storyDoc(storyID, objectID);
+            const meta = await doc.get().then((r) => r.data()!.metaData);
+            const trending = (meta.views ?? 0) - (meta.lastIndexedViews ?? 0);
+            await doc.update({"metaData.lastIndexedViews": meta.views ?? 0});
+            await index.partialUpdateObject({objectID, trending});
+          }
+        },
+      });
+    });
+
+export const indexRandomExploration = functions
+    .region("europe-central2")
+    .pubsub.schedule("every 24 hours")
+    .onRun(async () => {
+      await index.browseObjects<StoryRecord>({
+        query: "",
+        attributesToRetrieve: [],
+        batch: async (batch) => {
```

Comment:
Again, feels like querying Firestore using Algolia, when you can just query Firestore directly

Reply:
FS group queries, unfortunately, provide limited multi-collection querying functionality at this moment, so we have to keep it this way.